File: bwl.Imaging/Bitmap/DisplayObject.vb
Imports System.Drawing
Imports System.Runtime.Serialization
<KnownType(GetType(Polygon))>
<KnownType(GetType(Tetragon))>
<KnownType(GetType(Line))>
<KnownType(GetType(PointC))>
<KnownType(GetType(PointF))>
<KnownType(GetType(Point))>
<KnownType(GetType(RectangleF))>
<KnownType(GetType(Rectangle))>
<KnownType(GetType(TextObject))>
<KnownType(GetType(BitmapObject))>
<KnownType(GetType(Bitmap))>
Public Class DisplayObject
    Implements ICloneable
    Public Property Color As Color = Drawing.Color.Red
    Public Property IsMoveable As Boolean = False
    Public Property IsVisible As Boolean = True
    Public Property IsSelectable As Boolean = True
    Public Property ID As String = ""
    Public Property DrawObject As Object
    Public Property LineWidth As Single = 0.0
    Public Property PointSize As Single = 0.0
    Public Property Caption As String = ""
    Public Property Group As String = ""
    Public Property Parameters As New ParametersDictionaryList

    Public Sub New()
    End Sub

    Public Sub New(id As String, color As Color, obj As Object)
        Me.ID = id
        Me.Color = color
        Me.DrawObject = obj
    End Sub

    Public Sub New(id As String, color As Color, obj As Object, isSelectable As Boolean, isMoveable As Boolean)
        Me.ID = id
        Me.Color = color
        Me.DrawObject = obj
        Me.IsMoveable = isMoveable
        Me.IsSelectable = isSelectable
    End Sub

    Public Sub New(obj As DisplayObject)
        Me.Color = obj.Color
        Me.IsMoveable = obj.IsMoveable
        Me.IsVisible = obj.IsVisible
        Me.IsSelectable = obj.IsSelectable
        Me.ID = obj.ID
        Me.DrawObject = obj.DrawObject
        Me.LineWidth = obj.LineWidth
        Me.PointSize = obj.PointSize
        Me.Caption = obj.Caption
        Me.Group = obj.Group
    End Sub

    Public Function Clone() As Object Implements ICloneable.Clone
        Return New DisplayObject(Me)
    End Function
End Class

File: Bwl.Imaging.Skia/SKBitmap/DisplayObject.cs
using System.Runtime.Serialization;
using SkiaSharp;

namespace Bwl.Imaging.Skia;

[KnownType(typeof(Polygon))]
[KnownType(typeof(Tetragon))]
[KnownType(typeof(Line))]
[KnownType(typeof(SKPointC))]
[KnownType(typeof(SKPoint))]
[KnownType(typeof(SKPointI))]
[KnownType(typeof(SKRect))]
[KnownType(typeof(SKRectI))]
[KnownType(typeof(TextObject))]
[KnownType(typeof(SKBitmapObject))]
[KnownType(typeof(SKBitmap))]
public class DisplayObject : ICloneable
{
    public SKColor Color { get; set; } = SKColors.Red;
    public bool IsMoveable { get; set; } = false;
    public bool IsVisible { get; set; } = true;
    public bool IsSelectable { get; set; } = true;
    public string ID { get; set; } = "";
    public object DrawObject { get; set; }
    public float LineWidth { get; set; } = 0.0f;
    public float PointSize { get; set; } = 0.0f;
    public string Caption { get; set; } = "";
    public string Group { get; set; } = "";
    public ParametersDictionaryList Parameters { get; set; } = new ParametersDictionaryList();

    public DisplayObject()
    {
    }

    public DisplayObject(string id, SKColor color, object obj)
    {
        ID = id;
        Color = color;
        DrawObject = obj;
    }

    public DisplayObject(string id, SKColor color, object obj, bool isSelectable, bool isMoveable)
    {
        ID = id;
        Color = color;
        DrawObject = obj;
        IsMoveable = isMoveable;
        IsSelectable = isSelectable;
    }

    public DisplayObject(DisplayObject obj)
    {
        Color = obj.Color;
        IsMoveable = obj.IsMoveable;
        IsVisible = obj.IsVisible;
        IsSelectable = obj.IsSelectable;
        ID = obj.ID;
        DrawObject = obj.DrawObject;
        LineWidth = obj.LineWidth;
        PointSize = obj.PointSize;
        Caption = obj.Caption;
        Group = obj.Group;
    }

    public object Clone()
    {
        return new DisplayObject(this);
    }
}